pubsub/lib/pubsub_samples.ex

```diff
@@ -49,6 +75,173 @@ defmodule GoogleApi.PubSub.Samples do
+      conn,
+      project_id,
+      subscription_name,
+      [body: %{
+        topic: "projects/#{project_id}/topics/#{topic_name}"
+      }]
+    )
+
+    IO.puts "created subscription #{response.name} on topic #{response.name}"
+  end
+
+  @doc """
+  Publish a message to a PubSub topic.
+
+  ## Examples
+
+      iex> GoogleApi.PubSub.Samples.publish("YOUR_PROJECT_ID", "test-topic", "This is a message")
+      "published message 159505488737289"
+
+  """
+  def publish(project_id, topic_name, message) do
+    # Authenticate
+    {:ok, token} = Goth.Token.for_scope("https://www.googleapis.com/auth/cloud-platform")
+    conn = GoogleApi.PubSub.V1.Connection.new(token.token)
+
+    # Build the PublishRequest struct
+    request = %GoogleApi.PubSub.V1.Model.PublishRequest{
+      messages: [
+        %GoogleApi.PubSub.V1.Model.PubsubMessage{
+          data: Base.encode64(message)
+        }
+      ]
+    }
+
+    # Make the API request.
+    {:ok, response} = GoogleApi.PubSub.V1.Api.Projects.pubsub_projects_topics_publish(
+      conn,
+      project_id,
+      topic_name,
+      [body: request]
+    )
+
+    "published message #{response.messageIds}"
+    |> (&IO.ANSI.format([:green, :bright, &1])).()
+    |> IO.puts
+  end
+
+  @doc """
+  Runs a simple example which executes a pubsub supervisor and publishes to a
+  PubSub topic.
+
+  ## Examples
+
+      iex> GoogleApi.PubSub.Samples.run
+
+  """
+  def run() do
+    {:ok, _started} = Application.ensure_all_started(:goth)
+    :httpc.set_options(pipeline_timeout: 1000)
+    project_id = get_response("Project ID?: ")
+    topic_name = get_response("Topic Name?: ")
+    topic = get_or_create_topic(project_id, topic_name)
+    if topic do
+      subscription_name = get_response "Subscription Name?: "
+      subscription = get_or_create_subscription(project_id, topic_name, subscription_name)
+      if subscription do
+        {:ok, _pid} = start_subscription_supervisor(project_id, subscription_name)
+        get_next_message(project_id, topic_name)
+      end
+    end
+  end
+
+  defp start_subscription_supervisor(project_id, subscription_name) do
+    import Supervisor.Spec
+
+    # Define workers and child supervisors to be supervised
+    children = [
+      # Start the worker when the application starts
+      supervisor(GoogleApi.PubSub.Samples.Subscriber, [project_id, subscription_name]),
+    ]
+
+    # See https://hexdocs.pm/elixir/Supervisor.html
+    # for other strategies and supported options
+    opts = [strategy: :one_for_one, name: GoogleApi.PubSub.Samples.Supervisor]
+    Supervisor.start_link(children, opts)
+  end
+
+  defp get_or_create_topic(project_id, topic_name) do
+    GoogleApi.PubSub.Samples.get_topic(project_id, topic_name) ||
+      (confirm_topic_creation() && GoogleApi.PubSub.Samples.create_topic(project_id, topic_name))
+  end
+
+  defp confirm_topic_creation do
+    get_response("Topic does not exist. Create it? (y/n): ")
+    |> case do
```

Comment:
Now that we've moved most of the pipeline into a helper method, this is better:

```elixir
case get_response(...) do
...
```

same in the other confirm function.